File: apps/web/src/shared/api/rooms.ts
import { apiRequest } from "./client";

export type RoomVisibility = "public" | "private";

export type RoomSummary = {
  id: string;
  name: string;
  description: string | null;
  visibility: RoomVisibility;
  owner_user_id: string;
  member_count: number;
  is_member: boolean;
  is_owner: boolean;
  is_admin: boolean;
  is_banned: boolean;
  can_join: boolean;
  can_leave: boolean;
  can_manage_members: boolean;
  joined_at: string | null;
};

export type RoomMember = {
  id: string;
  username: string;
  joined_at: string;
  is_owner: boolean;
  is_admin: boolean;
  can_remove: boolean;
};

export type RoomBan = {
  id: string;
  user_id: string;
  username: string;
  banned_at: string;
  banned_by_username: string | null;
  reason: string | null;
};

export type RoomInvitation = {
  id: string;
  room_conversation_id: string;
  room_name: string;
  room_description: string | null;
  inviter_username: string | null;
  status: "pending" | "accepted" | "declined" | "revoked";
  created_at: string;
};

export type RoomListResponse = {
  rooms: RoomSummary[];
};

export type RoomInvitationListResponse = {
  invitations: RoomInvitation[];
};

export type RoomMemberListResponse = {
  members: RoomMember[];
};

export type RoomBanListResponse = {
  bans: RoomBan[];
};

export type CreateRoomPayload = {
  name: string;
  description?: string;
  visibility: RoomVisibility;
};

export type CreateInvitationPayload = {
  username: string;
  message?: string;
};

export const roomsApi = {
  listMine: () => apiRequest<RoomListResponse>("/api/rooms/mine"),
  listPublic: (search?: string) =>
    apiRequest<RoomListResponse>("/api/rooms/public", {
      query: search ? { search } : undefined,
    }),
  create: (payload: CreateRoomPayload) =>
    apiRequest<RoomSummary>("/api/rooms", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  join: (roomId: string) =>
    apiRequest<RoomSummary>(`/api/rooms/${roomId}/join`, {
      method: "POST",
    }),
  leave: (roomId: string) =>
    apiRequest<{ success: boolean; message: string }>(`/api/rooms/${roomId}/leave`, {
      method: "POST",
    }),
  listInvitations: () => apiRequest<RoomInvitationListResponse>("/api/rooms/invitations/mine"),
  invite: (roomId: string, payload: CreateInvitationPayload) =>
    apiRequest<RoomInvitation>(`/api/rooms/${roomId}/invitations`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  listMembers: (roomId: string) =>
    apiRequest<RoomMemberListResponse>(`/api/rooms/${roomId}/members`),
  listBans: (roomId: string) =>
    apiRequest<RoomBanListResponse>(`/api/rooms/${roomId}/bans`),
  removeMember: (roomId: string, memberUserId: string) =>
    apiRequest<{ success: boolean; message: string }>(
      `/api/rooms/${roomId}/members/${memberUserId}`,
      {
        method: "DELETE",
      },
    ),
  acceptInvitation: (invitationId: string) =>
    apiRequest<RoomSummary>(`/api/rooms/invitations/${invitationId}/accept`, {
      method: "POST",
    }),
};

File: apps/api/app/api/schemas/rooms.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints

from app.db.models.enums import InvitationStatus, RoomVisibility

RoomName = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True, min_length=3, max_length=120, pattern=r"^[A-Za-z0-9_.-]+$"
    ),
]
RoomDescription = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=500),
]
InvitationMessage = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=500)
]


class RoomSummaryResponse(BaseModel):
    id: UUID
    name: str
    description: str | None
    visibility: RoomVisibility
    owner_user_id: UUID
    member_count: int
    is_member: bool
    is_owner: bool
    is_banned: bool = False
    can_join: bool
    can_leave: bool
    joined_at: datetime | None = None


class RoomListResponse(BaseModel):
    rooms: list[RoomSummaryResponse]


class CreateRoomRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "engineering-room",
                "description": "Room for product and engineering coordination.",
                "visibility": "public",
            }
        }
    )

    name: RoomName
    description: RoomDescription | None = None
    visibility: RoomVisibility


class RoomInvitationResponse(BaseModel):
    id: UUID
    room_conversation_id: UUID
    room_name: str
    room_description: str | None
    inviter_username: str | None
    status: InvitationStatus
    created_at: datetime


class RoomInvitationListResponse(BaseModel):
    invitations: list[RoomInvitationResponse]


class CreateRoomInvitationRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "username": "alice",
                "message": "Join the private planning room for the launch workstream.",
            }
        }
    )

    username: Annotated[
        str,
        StringConstraints(
            strip_whitespace=True, min_length=3, max_length=64, pattern=r"^[A-Za-z0-9_.-]+$"
        ),
    ]
    message: InvitationMessage | None = None
